File: jambeth/jAmbeth.Cache.Bytecode/src/de/osthus/ambeth/bytecode/behavior/EmbeddedTypeBehavior.java
package de.osthus.ambeth.bytecode.behavior;

import java.util.List;

import de.osthus.ambeth.bytecode.EmbeddedEnhancementHint;
import de.osthus.ambeth.bytecode.visitor.EmbeddedTypeVisitor;
import de.osthus.ambeth.bytecode.visitor.InterfaceAdder;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.model.IEmbeddedType;
import de.osthus.ambeth.repackaged.org.objectweb.asm.ClassVisitor;

public class EmbeddedTypeBehavior extends AbstractBehavior
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Override
	public Class<?>[] getEnhancements()
	{
		return new Class<?>[] { IEmbeddedType.class };
	}

	@Override
	public ClassVisitor extend(ClassVisitor visitor, IBytecodeBehaviorState state, List<IBytecodeBehavior> remainingPendingBehaviors,
			List<IBytecodeBehavior> cascadePendingBehaviors)
	{
		if (state.getContext(EmbeddedEnhancementHint.class) == null)
		{
			return visitor;
		}
		visitor = new InterfaceAdder(visitor, IEmbeddedType.class);
		visitor = new EmbeddedTypeVisitor(visitor);
		return visitor;
	}
}

File: ambeth/Ambeth.Cache.Bytecode/ambeth/bytecode/behavior/EmbeddedTypeBehavior.cs
using De.Osthus.Ambeth.Bytecode.Visitor;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Model;
using System;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Bytecode.Behavior
{
    public class EmbeddedTypeBehavior : AbstractBehavior
    {
        [LogInstance]
        public ILogger Log { private get; set; }

        public override Type[] GetEnhancements()
        {
            return new Type[] { typeof(IEmbeddedType) };
        }

        public override IClassVisitor Extend(IClassVisitor visitor, IBytecodeBehaviorState state, IList<IBytecodeBehavior> remainingPendingBehaviors,
                    IList<IBytecodeBehavior> cascadePendingBehaviors)
        {
            if (state.GetContext<EmbeddedEnhancementHint>() == null)
            {
                return visitor;
            }
            visitor = new InterfaceAdder(visitor, typeof(IEmbeddedType));
            visitor = new EmbeddedTypeVisitor(visitor);
            return visitor;
        }
    }
}